HDDS-2981 Add unit tests for Proto [de]serialization

llemec opened the pull request:
## What changes were proposed in this pull request?

Added TestInstanceHelper to produce OzoneManagerProtocolProtos.PrefixInfo for testing purposes
Added test for getFromProtobuf and getProtobuf methods for OmPrefixInfo

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-2981

## How was this patch tested?
Unit tests

fapifta commented:
Hi @llemec,

thank you for working on this patch, I think overall the direction is good here, though I miss a few things, and I have a stylistic problem with the proposed changes.

In test testgetFromProtobufOneMetadataOneAcl(), we do check only the prefix path, metadata size and value, and acl size, should we also check for the acl value here? I think we should, if not, can you give some reasons why not?
Also, can you pleae remove the two empty lines at the end of this test method?

In test testGetProtobuf() we are also checking just for the count of object, should we check for the contents as well?

What I miss is some corner cases, as it seems, the code is prepared for metadata being null, and acl list being null, but the tests does not cover that scenario, so probably we should do so as well.

The stylistic problem I have is minor, and probably it is the matter of taste, I don't see the utility in having the TestInstanceHelper class, as at this point, that contains methods only for this test. In case these can be and should be used elsewhere we probably should move it out to a utility class at that time, if we notice it. If we miss it next time, I am more on the side of having duplications in the test code in well named methods, than having utility classes that might not get used ever elsewhere or having a too general name and purpose, while in the meantime incur some maintenance and cognitive costs whenever someone looks at them, so I prefer to put these kind of methods as private methods in my test class. If I convinced you, you can move it, or leave it as it is, up to you, I can accept this approach as well.

llemec commented:
Hello @fapifta,

Thank you for reviewing the diff. Addressing your points:

The testgetFromProtobufOneMetadataOneAcl() indeed only checks the presence of metadata and acl because looking at OmPrefixInfo#getFromProtobuf(), those parts of the message are included by calling the KeyValueUtil#getFromProtobuf() and OzoneAclUtil#fromProtobuf() methods respectively. To test the contents of the Acls and Metadata would mean in fact to test these methods implicitly. I propose we do those explicitly and in a separate jira.
By the same token with testGetProtobuf(), I propose we test the getProtobuf methods of KeyValueUtil and OzoneAclUtil separately and explicitly. 
In the explicit tests I agree, we should check the corner cases for both metadata and acls
Following this train of thought it is visible that Protobuf objects will be necessary for testing the different protos in Om. That is the proposed role of TestInstanceHelper - true, at the moment it only contains the helper methods for OmPrefixInfo testing.

Please let me know your thoughts on this.

Thank you,

ps: empty lines removal: sure.

fapifta commented:
Hello @llemec 

thank you for your comments and continued work on this PR. Indeed, I can accept the way based on your argument.

+1(non-binding) to merge the changes. Let's wait a committer to get it reviewed once more and committed if no further comments ;)

cxorm reviewed (commented):
Thanks @llemec for the work, I like the explanation in the above review.
I have some nits commented inline.

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestInstanceHelper.java:
```diff
@@ -0,0 +1,67 @@
+ */
+public final class TestInstanceHelper {
+
+  private  TestInstanceHelper(){
```
done

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestOmPrefixInfo.java:
```diff
@@ -18,39 +18,91 @@
+      IAccessAuthorizer.ACLIdentityType identityType,
+      String identityString,
+      IAccessAuthorizer.ACLType aclType,
+      OzoneAcl.AclScope scope){
```
done

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestOmPrefixInfo.java:
```diff
@@ -18,39 +18,91 @@
   }
+
+  @Test
+  public void testgetFromProtobufOneMetadataOneAcl(){
```
done

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestInstanceHelper.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  public static OzoneManagerProtocolProtos.PrefixInfo getDefaultTestPrefixInfo(
+      String name, String aclString, HddsProtos.KeyValue metadata){
```
done

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestInstanceHelper.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  public static HddsProtos.KeyValue getDefaultTestMetadata(
+      String key, String value){
```
done

llemec commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestInstanceHelper.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  public static OzoneManagerProtocolProtos.OzoneAclInfo buildTestOzoneAclInfo(
+      String aclString){
```
done

llemec commented:
Hello @cxorm,

Thank you for your review. I corrected the coding style as suggested. In a larger perspective, would it make sense to enhance checkstyle rules to warn for this kind of spacing?

Thanks,

cxorm commented:
>In a larger perspective, would it make sense to enhance checkstyle rules to warn for this kind of spacing?

Yeah, thanks for the feedback, we could discuss about the idea offline : )

cxorm commented:
The work looks great to me.

A little question is about the title of the issue,
I think the title is too general, could we change it to address the fix of this PR ?
Feel free to let me know your thoughts, thanks : )

adoroszlai commented on hadoop-ozone/common/src/test/java/org/apache/hadoop/ozone/om/helpers/TestOmPrefixInfo.java:
```diff
@@ -18,39 +18,91 @@
-  public void testCopyObject() {
-    OmPrefixInfo omPrefixInfo = new OmPrefixInfo("/path",
+  public OmPrefixInfo getOmPrefixInfoForTest(String path,
+      IAccessAuthorizer.ACLIdentityType identityType,
+      String identityString,
+      IAccessAuthorizer.ACLType aclType,
+      OzoneAcl.AclScope scope) {
```
It could be improved by accepting an `OzoneAcl` instead of a group ACL-related parameters.

adoroszlai reviewed (changes requested):
Thanks @llemec for adding these tests.  I'd like to suggest some refactoring to reduce the amount of duplicated code and make it easier to follow the tests.

adoroszlai commented:
Thanks @llemec for the patch.  On second thought I think it's fine to commit as is and refactor later.  I have merged current master into the branch and resolved some conflicts caused by recent code reorg.

adoroszlai commented:
@fapifta @cxorm would you like to review the latest update (where I merged from `master`)?

cxorm commented:
Sorry for missing it, thank you @adoroszlai for updating it and resolving the conflict,
I'm +1.

adoroszlai commented:
Thanks @llemec for the contribution, and sorry for taking so long to merge it.  Thanks @cxorm and @fapifta for the review.
